internal/uvm/vpmem.go

```diff
@@ -2,57 +2,195 @@ package uvm
 
 import (
 	"context"
-	"errors"
 	"fmt"
 	"os"
 
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+
 	"github.com/Microsoft/hcsshim/internal/guestrequest"
 	"github.com/Microsoft/hcsshim/internal/log"
 	"github.com/Microsoft/hcsshim/internal/requesttype"
 	hcsschema "github.com/Microsoft/hcsshim/internal/schema2"
-	"github.com/sirupsen/logrus"
 )
 
 const (
-	lcowVPMEMLayerFmt = "/run/layers/p%d"
+	lcowVPMEMLayerFmt        = "/run/layers/p%d"
+	lcowVPMEMStackedLayerFmt = "/run/layers/p%d-%d-%d"
+	PageSize                 = 0x1000
 )
 
 var (
 	// ErrMaxVPMEMLayerSize is the error returned when the size of `hostPath` is
 	// greater than the max vPMEM layer size set at create time.
 	ErrMaxVPMEMLayerSize = errors.New("layer size is to large for VPMEM max size")
+	ErrNotEnoughSpace    = errors.New("not enough space")
 )
 
+type slotDescriptor struct {
+	deviceNumber uint32
+	slotPosition int
```

Comment:
IMO this struct isn't really necessary. I think maybe something like the below would be more straightforward:

```
type vpmemMapping struct {
	hostPath     string
	uvmPath      string
	deviceOffset uint64
	deviceSize   uint64
	refCount     uint32
}

type vpmemDevice struct {
	maxSize  uint64
	deviceNumber uint32
	mappings []*vpmemMapping
}
```

and then just return and pass around the target slot position as needed. 

What do you think?

Reply:
yup, I ended up removing it.